targets/psa_switch/psa_switch.cpp

```diff
@@ -354,31 +372,42 @@ PsaSwitch::ingress_thread() {
     Deparser *deparser = this->get_deparser("ingress_deparser");
     deparser->deparse(packet.get());
 
-    // handling multicast
-    unsigned int mgid = 0u;
-    const auto &f_mgid = phv->get_field("psa_ingress_output_metadata.multicast_group");
-    mgid = f_mgid.get_uint();
-
+    // multicast
+    unsigned int mgid = phv->get_field("psa_ingress_output_metadata.multicast_group").get_uint();
     if(mgid != 0){
-      BMLOG_DEBUG_PKT(*packet,
-                      "Multicast requested for packet with multicast group {}",
-                      mgid);
-      const auto pre_out = pre->replicate({mgid});
-      auto packet_size = packet->get_register(PACKET_LENGTH_REG_IDX);
-      for(const auto &out : pre_out){
-        auto egress_port = out.egress_port;
-        BMLOG_DEBUG_PKT(*packet, "Replicating packet on port {}", egress_port);
-        std::unique_ptr<Packet> packet_copy = packet->clone_with_phv_ptr();
-        packet_copy->set_register(PACKET_LENGTH_REG_IDX, packet_size);
-        enqueue(egress_port, std::move(packet_copy));
-      }
+      multicast(packet.get(), mgid);
       continue;
     }
+    // TODO(pete) from psa spec - else if multicast group is empty, drop packet
```

Comment:
I am not sure I understand this comment. Why would you drop the packet is the mcast group is empty?

Reply:
Ah thanks - I had forgotten to ask @jafingerhut about this. It's very possible I'm reading too far into the spec.

There's a paragraph under Packet Cloning in the PSA spec that says clone session configs should work similarly to multicast group configs.
Under Multicast replication, the first paragraph says "Sending a packet to an empty group causes the packet to
be dropped."
https://github.com/p4lang/p4-spec/blob/master/p4-16/psa/PSA.mdk#L899

@antoninbas I don't have a great answer to your question since it is also a question of mine.
@jafingerhut, can you shed some light on whether I'm reading too much into the spec by adding this TODO or if there's a scenario in which it is useful?

Reply:
If the multicast_group field is non-0, then it is possible that the multicast group has been configured with no (egress_port, egress_instance) pairs at all.  If so, then no copies of the packet are made, so it has a similar effect as dropping the packet, in that no packets are enqueued in the packet buffer for any output ports at all.

If a packet should be cloned, because the ostd.clone field is true when ingress is done, then the clone session configuration for clone session numbered ostd.clone_session_id.  That clone session might be configured with no (egress_port, egress_instance) pairs.  If that set of pairs is empty, then no clone packets are created.

Does that answer your question?

Reply:
Note that whether clones are created for a packet, is an independent decision from whether the packet is dropped or not.  See the pseudocode for end-of-ingress behavior in section https://p4.org/p4-spec/docs/PSA-v1.1.0.html#sec-after-ingress

and for end-of-egress behavior in section https://p4.org/p4-spec/docs/PSA-v1.1.0.html#sec-after-egress

Feel free to ask any questions about those or any other parts of the PSA spec they are not clear.

Reply:
That seems pretty easy to implement, we should have the behavior that @jafingerhut is describing from the get go.

Reply:
Specifically, the current code already implements the correct behavior for empty multicast groups, so the TODO should be removed. However, it seems that more work is required to ensure that the cloning decision can be enforced even when the packet is dropped (that includes when the multicast group is empty).

Reply:
Thanks @jafingerhut and @antoninbas, taking a look at this later today as well.